tasks/plugins.yml

```diff
@@ -6,15 +6,23 @@
     paths: "{{ grafana_data_dir }}/plugins"
   register: installed_plugins
 
-- name: Install plugins
+- name: Remove unwanted plugins
   become: true
-  command: "grafana-cli --pluginsDir {{ grafana_data_dir }}/plugins plugins install {{ item }}"
-  args:
-    creates: "{{ grafana_data_dir }}/plugins/{{ item }}"
-  with_items: "{{ grafana_plugins | difference(installed_plugins.files) }}"
-  register: _plugin_install
-  until: _plugin_install is succeeded
-  retries: 5
-  delay: 2
+  grafana_plugin:
+    grafana_plugins_dir: "{{ grafana_data_dir }}/plugins"
+    name: "{{ item.path | basename }}"
+    state: absent
+  when: item.path | basename  not in grafana_plugins
+  with_items: "{{ installed_plugins.files }}"
+  notify:
+    - restart grafana
+
+- name: Make sure plugins are installed and updated
+  become: true
+  grafana_plugin:
+    grafana_plugins_dir: "{{ grafana_data_dir }}/plugins"
+    name: "{{ item }}"
+    state: present
+  with_items: "{{ grafana_plugins }}"
```

Comment:
What do you think about merging those two tasks together and modifying `grafana_plugins` into a map? For now we could have sth like:
```
grafana_plugins:
- name: <plugin_name>
  state: "installed"
```

This way it provides some sort of immutability and we don't need to care about detecting `installed_plugins` as this is something module already can do.

Reply:
I don't like dealing with ansible modules, which doesn't support declarative way of configuration of resources neither. I would rather have some option to specify that plugins list is exclusive, like it does module authorized_key. In my opinion the solution which we have now is best way to make it declarative. If we add `state` key to variable `grafana_plugins` we loose the declarative way of configuration. 

By the way this is what i meant by breaking changes, I didn't wanted to force users of this role to make changes in vars. If it's ok for you i would add another (optional) key `version` to be able to choose  specific version.

Anyway if you want to simplify this code, I would rather make issue in this module to support feature I mentioned earlier. Without support in module this can't be done in a nice way.

Reply:
It was already decided that 0.17.0 will be the last version that is compatible with previous ones. This is due to the fact that migration to grafana modules (instead of using `uri` module) is going to break some things and such move is allowed as per sem-ver definitions.
There is a `release-0.17.0` branch cut off if we need to make some bugfix changes, but all features are going into `master`. 

---
> I don't like dealing with ansible modules, which doesn't support declarative way of configuration of resource

Every module which exposes `state` option and allows passing `present` or `absent` is declarative by nature. Problem here is "what" needs to be declarative, in case of majority of ansible modules it is declaration of one resource (key/plugin/whatever) and not a list of resources. This is also true for `authorized_key` default behavior where `exclusive` is set to false.
One of the reasons we are migrating to grafana modules is to solve such issues in the long run as it is much easier to add support for `exlusive` in python code, than doing a lot of workarounds in ansible role.

---
I get that you want to have a way to declare a list of plugins and make it immutable. However such feature should be supported inside `grafana_plugin` module and not externally (your example of `authorized_key` is perfect here). Otherwise, it just increases already complicated ansible role. Maybe consider adding it to https://github.com/ansible-collections/community.grafana/blob/master/plugins/modules/grafana_plugin.py ?

Reply:
> It was already decided that 0.17.0 will be the last version that is compatible with previous ones. This is due to the fact that migration to grafana modules (instead of using uri module) is going to break some things and such move is allowed as per sem-ver definitions.
There is a release-0.17.0 branch cut off if we need to make some bugfix changes, but all features are going into master.

I'm glad to hear that we can do breaking changes, this allow to make it better in long run.

>I get that you want to have a way to declare a list of plugins and make it immutable. However such feature should be supported inside grafana_plugin module and not externally (your example of authorized_key is perfect here). Otherwise, it just increases already complicated ansible role. Maybe consider adding it to https://github.com/ansible-collections/community.grafana/blob/master/plugins/modules/grafana_plugin.py ?

I can make PR to grafana_plugin module, however I would like to have quick fix until it's merged. I don't have experiences with dynamic of community.grafana project, so I can't predict when it will be available in ansible.

Reply:
> dynamic of community.grafana project, so I can't predict when it will be available in ansible.

It is currently used as an experiment to break down ansible monolithic repository into smaller chunks, so changes are adopted pretty fast.

---
> however I would like to have quick fix until it's merged

Ok, let's move forward with this, but please add some reference in code to issue/PR in community.grafana project, so we can remove this hack when ansible module has this functionality.